ECDC-4533: Multiple plots and only one consumer in Daqlite

madsipsen opened the pull request:
Jira ticket: https://jira.ess.eu/browse/ECDC-4533

### Overview
This PR is a refactorization of the daqlite code, which enables

- Display of multiple plots by running only one instance of daqlite. Previously, a separate instance was required for each plot. This is handled by introducing the key `plots` in the config file, where several `plot` keys then can be specified.
- Functionality for handling multiple plots is handled by new methods added to the `Configuration` class.
- Only one instance of the `WorkerThread` and the `ESSConsumer` is used for handling the communication of Kafka topics to the different plots
- Each plot class must call `ESSConsumer::addSubscriber`. The `ESSConsumer` then uses the three private containers
  - `mSubscribers`
  - `mSubscriptionCount`
  - `mDeliveryCount`
- These containers are used for bookkeeping the number of plot subscribers, the type of data a given plot requires, and keep track of the data count delivered so far. Previously all data was cleared after the first call to the `ESSConsumer` query methods `readResetHistogram`, `readResetHistogramTof`, etc. Now the data is first cleared when the subscription of the last plot has been satisfied.

Examples of multi-plot config files can be found here

- configs/amor/multi_amor.json
- configs/loki/multi_loki.json

### Refactorization
The follow code clean-up has also been done
- Use of modern style Qt connections (invalid connections will no longer compile)
- [IWYU](https://github.com/include-what-you-use/include-what-you-use) driven header clean-up. This spots situations where forward declarations can be used, in order to lower the header include overhead.
- A lot of member data has been made private, and is now be accessed using appropriate getter methods

### Todo
- Remove all the subscription booking from the `ESSConsumer`, and add this to a new class as a separation layer/interface between the ESSConsumer and the plots. 
- Specify `tof` and `geometry` options for each plot. Currently these can only be given as top-level options shared by all plots.
- Improve GUI with better layout and improved usability

Commit:
- Working stuff

Commit:
- ESSConsumer handling of data for multiple plots in place

Commit:
- Rename variables used for member data

Commit:
- Add multi-plot IO handling to the Configuration class

Commit:
- Modern style Qt connections

Commit:
- Use IWYU to clean-up header includes

Commit:
- Connection clean-up

Commit:
- Minor clean-up

Commit:
- Fixing build issue

mortenjc commented on src/daqlite/AbstractPlot.h:
```diff
@@ -8,28 +8,33 @@
-      : mPlotType(Type), mConsumer(Consumer) {}
+  : mPlotType(Type)
+  , mConsumer(Consumer) 
+{
```
Bruger parentes 'at end of line' og ikke stand alone.

madsipsen replied:
Fixed

madsipsen replied:
Done

mortenjc commented on src/daqlite/Configuration.cpp:
```diff
@@ -5,17 +5,98 @@
-void Configuration::fromJsonFile(std::string fname) {
+#include "Configuration.h"
+
+#include <nlohmann/detail/iterators/iter_impl.hpp>
```
are  these includes necessary? They don't look like thgey were intended for direct inclusion

madsipsen replied:
Fixed

madsipsen replied:
Done

mortenjc commented on src/daqlite/Configuration.cpp:
```diff
@@ -5,17 +5,98 @@
+#include <nlohmann/detail/iterators/iteration_proxy.hpp>
+#include <nlohmann/json.hpp>
+
+#include <fmt/core.h>
```
we normally use fmt/fmt.h what is the difference?

madsipsen replied:
There is a good description [here](https://github.com/fmtlib/fmt/issues/1755#issuecomment-654525299). `<fmt/core.h>` is more lightweight whereas `<fmt/format.h>` pulls in the whole API. So for this "modest" use, I think `<fmt/core.h>` is the correct choice.

mortenjc commented on src/daqlite/Configuration.h:
```diff
@@ -10,19 +10,31 @@
 #pragma once
 
-#include <fstream>
+#include <nlohmann/json_fwd.hpp>
```
again, I am not used to see these nlohmann includes, does the 'normal' inlcude not work?

madsipsen replied:
Done

mortenjc commented on src/daqlite/Custom2DPlot.cpp:
```diff
@@ -5,24 +5,30 @@
 
-#include <Custom2DPlot.h>
-#include <WorkerThread.h>
+#include "Custom2DPlot.h"
```
We use the <> includes generally in the efu and daqlite projects.

madsipsen replied:
Done

mortenjc commented on src/daqlite/Custom2DPlot.cpp:
```diff
@@ -5,24 +5,30 @@
+    : AbstractPlot(PlotType::PIXEL, Consumer)
+    , mConfig(Config)
+    , mProjection(Proj)
+{
```
mProjection(Proj)  {
...

madsipsen replied:
Done

mortenjc commented on src/daqlite/CustomAMOR2DTOFPlot.cpp:
```diff
@@ -5,30 +5,38 @@
-#include <algorithm>
-#include <assert.h>
-#include <fmt/format.h>
+#include "CustomAMOR2DTOFPlot.h"
```
<>

madsipsen replied:
Done

mortenjc reviewed (changes requested):
Ser fint ud. Der er lidt med includes hvor jeg er i tvivl om du optimerer eller om ikke man i stedet kan bruge en mere generisk include.

og helt generelt har vi valgt for længe siden udelukkende at bruge <> i includes.

Commit:
- Streamlining headers

Commit:
- Updating the copyright year interval

Commit:
- Introducing "using std::string" etc. to lower the C++ verbosity
- More clean-up

Commit:
- Fixing "{" in ctors

Commit:
- Using <fmt/format.h>

Commit:
- Fixing one more "{" in ctors

tbukovics commented on src/daqlite/Configuration.cpp:
```diff
@@ -29,93 +109,93 @@ void Configuration::fromJsonFile(std::string fname) {
-  Plot.ColorGradient = getVal("plot", "color_gradient", Plot.ColorGradient);
-  Plot.InvertGradient = getVal("plot", "invert_gradient", Plot.InvertGradient);
-  Plot.LogScale = getVal("plot", "log_scale", Plot.LogScale);
+  mPlot.PlotType = getVal("plot", "plot_type", mPlot.PlotType);
```
You moved the enum PlotType into Common.h. It would be nice if that enum would be a bit more advanced (class wrapper), and would able to be constructed from the string which already sets the rgiht enum value, or throw if the string is not valid PlotType. This then can be reused many places and ensures we have a common string->enum relation

madsipsen replied:
Fixed

tbukovics commented on src/daqlite/MainWindow.cpp:
```diff
@@ -49,16 +57,18 @@ MainWindow::MainWindow(Configuration &Config, QWidget *parent)
-  } else if (strcmp(mConfig.Plot.PlotType.c_str(), "tof") == 0) {
+  }
+
+  else if (strcmp(mConfig.mPlot.PlotType.c_str(), "tof") == 0) {
```
We should encapsulate this into PlotType object which ensures string construction to the right internal enum value or throws. So we catch string typos already at conf file processing and we have one common point to transform ENUM to string.

madsipsen replied:
Fixed

tbukovics commented on src/daqlite/Common.h:
```diff
@@ -0,0 +1,18 @@
+
+#pragma once
+
+enum class PlotType {
```
Please rename this from Common.h to CommmonTypes.h and add a class wrapper arounf this enum, which ensures that this enum can be constructed from a string and it's values can be converted to a string. See my other comment regarding to this point.

Also I would create a similar string interpreter "enum" class for the ColorGradient to replace the string map in each 2D plot with one common enum.

madsipsen replied:
Done - and thanks for the nice suggestion!

tbukovics commented on src/daqlite/ESSConsumer.cpp:
```diff
@@ -329,3 +346,117 @@ std::unique_ptr<RdKafka::Message> ESSConsumer::consume() {
+vector<uint32_t> ESSConsumer::readResetTOFs() {
+  vector<uint32_t> ret = mTOFs;
+
+  if (checkDelivery(DataType::TOF)) {
```
I see now with the sibscribers you follow up how "claimed" the data from the consumer to block the reset until all plots done with consume data. This is a proper workaround for this minimal viable product, but it's clearly indicate we have to refactor how plots served with data becase it's not a healty design that they pull. They should get served by this ESSconsumer (which is a producer :D)

Nothing to be done here, just a comment.

madsipsen replied:
Thanks!

tbukovics commented on src/daqlite/Configuration.h:
```diff
@@ -72,7 +82,7 @@ class Configuration {
 
-  struct Plot {
+  struct PlotOptions {
     std::string PlotType{"pixels"}; // "tof" and "tof2d" are also possible
```
If you agree to upgrade a bit the PlotType enum object (my other commments) here we can already use that to store the proper type, so w can eliminate string values.

madsipsen replied:
Fixed

tbukovics commented on src/daqlite/CustomAMOR2DTOFPlot.cpp:
```diff
@@ -102,7 +112,7 @@ void CustomAMOR2DTOFPlot::setCustomParameters() {
 // provide a list of options
 QCPColorGradient
-CustomAMOR2DTOFPlot::getColorGradient(std::string GradientName) {
+CustomAMOR2DTOFPlot::getColorGradient(const std::string &GradientName) {
```
It would be nice to introduce a similar advanced enum kind of class which can to the trasformation btw. strings and QCPColorGradient types. This function should be merged into this object and that should be in the abstract plot. Curently we have this ugly string map for independently for each 2D plot with the same colors.

Also we can initialize this "enum" gradient type class when we read config.

madsipsen replied:
This is harder to do for `QCPColorGradient` which resides in the qcustomplot library

tbukovics reviewed (changes requested):
So this is a transition state, because the plot<->consumer relation is not healty at all, but I think you found a good workaround to trace subscription.

I generally ok with you modification, but I would upgrade a bit refactor for string parsing for plot types and color gradient to maintain all these transformations in one place.

Commit:
- Refactoring PlotType enum to more flexible class
- Fixing warning about unused variable
- std::string support for PlotType enum

tbukovics commented on src/daqlite/Configuration.h:
```diff
@@ -83,7 +85,7 @@ class Configuration {
 
   struct PlotOptions {
-    std::string PlotType{"pixels"}; // "tof" and "tof2d" are also possible
+    PlotType Plot{"pixels"}; // "tof" and "tof2d" are also possible
```
I would use the enum for initialize this not a string

madsipsen replied:
Fixed

Commit:
- Adding updated PlotType enum class
- Adding DataType enum class
- Adding types() method for looping through enums in an enum class
- Cleaning up ESSConsumer

Commit:
- Use PlotType explicitly in ctor call

mortenjc commented on src/daqlite/Configuration.h:
```diff
@@ -1,4 +1,4 @@
-// Copyright (C) 2020 - 2021 European Spallation Source, ERIC. See LICENSE file
+// Copyright \(C\) 2020 - 2025 European Spallation Source, ERIC. See LICENSE file
```
Probably not intended?

madsipsen replied:
Fixed

mortenjc commented on src/daqlite/Custom2DPlot.cpp:
```diff
@@ -1,28 +1,37 @@
-// Copyright (C) 2020 - 2021 European Spallation Source, ERIC. See LICENSE file
+// Copyright \(C\) 2020 - 2025 European Spallation Source, ERIC. See LICENSE file
```
result of regex replace?

madsipsen replied:
Fixed

mortenjc commented on src/daqlite/Custom2DPlot.h:
```diff
@@ -1,4 +1,4 @@
-// Copyright (C) 2020 - 2024 European Spallation Source, ERIC. See LICENSE file
+// Copyright \(C\) 2020 - 2025 European Spallation Source, ERIC. See LICENSE file
```
same

madsipsen replied:
Fixed

mortenjc commented on src/daqlite/CustomAMOR2DTOFPlot.cpp:
```diff
@@ -1,34 +1,44 @@
-// Copyright (C) 2020 - 2022 European Spallation Source, ERIC. See LICENSE file
+// Copyright \(C\) 2020 - 2025 European Spallation Source, ERIC. See LICENSE file
```
...

madsipsen replied:
Fixed

mortenjc commented on src/daqlite/CustomTofPlot.cpp:
```diff
@@ -1,31 +1,40 @@
-// Copyright (C) 2020 - 2021 European Spallation Source, ERIC. See LICENSE file
+// Copyright \(C\) 2020 - 2025 European Spallation Source, ERIC. See LICENSE file
```
...

madsipsen replied:
Fixed

mortenjc commented on src/daqlite/CustomTofPlot.h:
```diff
@@ -1,4 +1,4 @@
-// Copyright (C) 2020 - 2024 European Spallation Source, ERIC. See LICENSE file
+// Copyright \(C\) 2020 - 2025 European Spallation Source, ERIC. See LICENSE file
```
...

mortenjc commented on src/daqlite/ESSConsumer.h:
```diff
@@ -1,4 +1,4 @@
-// Copyright (C) 2020 - 2024 European Spallation Source, ERIC. See LICENSE file
+// Copyright \(C\) 2020 - 2025 European Spallation Source, ERIC. See LICENSE file
```
...

mortenjc commented on src/daqlite/HistogramPlot.cpp:
```diff
@@ -1,34 +1,45 @@
-// Copyright (C) 2020 - 2021 European Spallation Source, ERIC. See LICENSE file
+// Copyright \(C\) 2020 - 2025 European Spallation Source, ERIC. See LICENSE file
```
...

mortenjc commented on src/daqlite/HistogramPlot.h:
```diff
@@ -1,4 +1,4 @@
-// Copyright (C) 2020 - 2024 European Spallation Source, ERIC. See LICENSE file
+// Copyright \(C\) 2020 - 2025 European Spallation Source, ERIC. See LICENSE file
```
...

madsipsen replied:
Fixed

mortenjc commented on src/daqlite/MainWindow.cpp:
```diff
@@ -1,44 +1,52 @@
-// Copyright (C) 2020 - 2022 European Spallation Source, ERIC. See LICENSE file
+// Copyright \(C\) 2020 - 2025 European Spallation Source, ERIC. See LICENSE file
```
...

madsipsen replied:
Fixed

mortenjc commented on src/daqlite/MainWindow.h:
```diff
@@ -1,4 +1,4 @@
-// Copyright (C) 2020 - 2024 European Spallation Source, ERIC. See LICENSE file
+// Copyright \(C\) 2020 - 2025 European Spallation Source, ERIC. See LICENSE file
```
...

madsipsen replied:
Fixed

mortenjc commented on src/daqlite/WorkerThread.h:
```diff
@@ -1,4 +1,4 @@
-// Copyright (C) 2020 - 2024 European Spallation Source, ERIC. See LICENSE file
+// Copyright \(C\) 2020 - 2025 European Spallation Source, ERIC. See LICENSE file
```
...

madsipsen replied:
Fixed

mortenjc commented on src/daqlite/types/DataType.h:
```diff
@@ -0,0 +1,149 @@
+  };
+
+  // Max and min enum values
+  static constexpr int MIN = Types::NONE;
```
Ok, so this idea is really no much better than using a macro in this respect if min and max have to be specified manually anyway :-)

madsipsen replied:
Fixed

mortenjc reviewed (commented):
I believe most of this was reviewed earlier?

Some weird \(C\) has entered the copyright header, probably as a result of regex substitution.
